feat: add tips for each module while processing speech to text

baptisteghostine opened the pull request:
#### **Overview**
This PR brings substantial improvements to the Speaking screen and its corresponding tests, focusing on enhancing the user experience during the "processing" state. When the app analyzes the user's responses (transcription), a semi-transparent overlay with context-specific tips and a progressive loading bar is displayed. Once the analysis is complete, the progress bar quickly fills, and the user is navigated back to the previous screen.

---

#### **Related Issues**
- **Issue #247 **
- **Issue #248 **
- **Issue #249 **

![image](https://github.com/user-attachments/assets/d904d2ea-8a96-4c8c-a24b-f01e9bbc2d6d)


---

#### **Key Changes**

##### **1. Processing Overlay with Tips and Progress Bar**
- A semi-transparent overlay appears during the `ANALYSIS_STATE.PROCESSING` phase.
- Context-specific tips are displayed based on the current practice module:
  - **Interview Context**: Tips on minimizing filler words and structuring answers.
  - **Public Speaking Context**: Suggestions for improving engagement techniques or storytelling.
  - **Sales Pitch Context**: Recommendations for emphasizing unique selling points or handling objections.
- A linear progress bar incrementally fills:
  - Starts slowly during processing.
  - Quickly completes when the analysis finishes before navigating the user back.

##### **2. Context-Aware Tips**
- Added `getTipsForModule(apiLinkViewModel)` function to generate context-specific tips.
- Displays a generic motivational tip if no context is set.

##### **3. Enhanced User Feedback During Processing**
- Clear visual indicators reassure users that their responses are being analyzed.
- The overlay and progress bar improve user experience by reducing uncertainty during the processing phase.

##### **4. Refined State Management**
- Maintains a logical flow of states:
  - `RECORDING` -> `PROCESSING` (with overlay) -> `FINISHED`.
  - Automatically navigates back after finishing.
- Ensures `endAndSave()` is called before transitioning to the next screen.

##### **5. Testing and Code Coverage**
- Added tests to verify:
  - Overlay, tips container, and progress bar visibility during the `PROCESSING` state.
  - Correct tips appear for different contexts (Interview, Public Speaking, Sales Pitch).
  - UI elements (mic button, back button, transcribed text, sentiment analysis text) remain stable across states.
- Increased line coverage by testing error states, partial states, and edge cases.

---

#### **Example Interaction**
1. **User finishes recording (`RECORDING -> PROCESSING`)**:
   - The screen dims and shows a tip:  
     ```plaintext
     Highlight both technical and soft skills relevant to the role.
     ```
   - A progress bar starts filling slowly.
2. **Processing completes**:
   - The progress bar quickly fills to 100%.
   - The user is navigated back to the previous screen.

---

#### **Future Enhancements**
- **Internationalization**: Add support for multiple languages to broaden accessibility.
- **Enhanced Visuals**: Include contextual animations or clues as the progress bar nears completion.
- **Diverse Tips**: Expand the pool of tips for each module to keep the experience engaging and fresh for repeat users.

---

#### **Conclusion**
This PR greatly enhances the user experience by adding visual feedback and helpful tips during the processing phase of speech analysis. It improves user engagement by providing contextually relevant advice and reassuring feedback, while also increasing test coverage and overall code robustness. A highly valuable and user-centric improvement to the app’s speaking practice flow.

ymarc11 reviewed (changes requested):
This PR introduces an impressive feature that enhances the "Speaking" screen by adding contextual tips, a semi-transparent overlay, and a progress bar during the processing state (while waiting after the recording). It effectively improves user engagement and enjoyment during speech analysis. The additional test coverage ensures robustness. Good code coverage, the code duplication is under 3% and the CI passes, well done. Also. your description is very clear.

Contextual Tips and User Feedback:

- Including tips for specific practice modules (ex: interview, public speaking, sales pitch) improves user experience and provides actionable and targeted advice.
- Adding general motivational tips when no context is set is a good idea.

Progress Bar Implementation:

- A progress bar gives visual feedback during the processing state, reducing user uncertainty which i would say would please the users.

Clean Code Structure:

- The getTipsForModule function is good and clear. It would be nice if we could have the option to add additional tips to the lists (but it is not very important for now, could do it later if we have time).

Good Test Coverage:

- The tests comprehensively cover the various states and ensure the correctness of displayed tips and UI elements across different contexts.
- Use of semantic nodes and test tags makes tests maintainable and robust.

Suggestions or Improvement:
- Your documentation can sometimes be a little bit more detailed (like for example the kdoc for getTipsForModule. But it still understandable, so not so much of a problem
- I think getTipsForModule could be put in the apiLinkViewModel instead of the Speaking screen to avoid mixing backend and frontend.
- (Optional): Could have added a function that lets us add on the list of tips

This PR is a well-rounded improvement to the "Speaking" screen, offering a meaningful, user friendly update. Other than the move of getTipsForModule to the view model, your implementation is ready to be merged.

ymarc11 commented on app/src/main/java/com/github/se/orator/ui/speaking/Speaking.kt:
```diff
@@ -67,67 +57,210 @@ fun SpeakingScreen(navigationActions: NavigationActions, viewModel: SpeakingView
+  // Example tips for each module
+  val interviewTips =
+      listOf(
+          "Avoid filler words like 'um' or 'uh' to sound more confident.",
```
Nicely implemented. I just think it would be better to add it in the ApiLinkViewModel instead of the Speaking screen. (And maybe a little more detail in the documentation).

baptisteghostine replied:
Yes I agree! Done

baptisteghostine commented on app/src/main/java/com/github/se/orator/ui/speaking/Speaking.kt:
```diff
@@ -67,67 +57,210 @@ fun SpeakingScreen(navigationActions: NavigationActions, viewModel: SpeakingView
+    is SalesPitchContext -> salesPitchTips
+    else -> {
+      // If no context or a different one is set, return a generic tip list
+      listOf("Stay calm, be clear, and remember your main goals!")
```
Thanks!

baptisteghostine commented:
Hello Marc, thank you for your great PR review. I took into consideration your comments and it's now fixed. 

I also removed hardcoded values by updating Theme.kt.

If everything works for you, the PR can be merged as there's no conflicts. 

Have a great evening!
